### PR TITLE
Perf: Improve Performance, reduce heap allocations

### PR DESCRIPTION
**Pull request by soujanyanmbri**
> Thank you for contributing to Coraza WAF, your effort is greatly appreciated
> Before submitting check if what you want to add to `coraza` list meets [quality standards](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#quality-standards) before sending pull request. Thanks!

**Make sure that you've checked the boxes below before you submit PR:**

- [x] My code includes positive and negative tests.
- [x] I have an appropriate description with correct grammar.
- [x] I have read [Contribution guidelines](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#contribution-guidelines), [maintainers note](https://github.com/corazawaf/coraza/blob/master/CONTRIBUTING.md#maintainers) and [Quality standard](https://github.com/github.com/corazawaf/coraza/v3sso/coraza-waf/blob/master/CONTRIBUTING.md#quality-standards).
- [x] My code is properly linted and passes pre-commit tests.

Thanks for your contribution :heart:

Before: 
```
BenchmarkTxSetGet/Set-10         	26881494	        42.25 ns/op	      32 B/op	       1 allocs/op
BenchmarkTxSetGet/Get-10         	15812412	        76.35 ns/op	      48 B/op	       2 allocs/op
```
After: 
```
BenchmarkTxSetGet/Set-10         	66049288	        18.02 ns/op	       0 B/op	       0 allocs/op
BenchmarkTxSetGet/Get-10         	25339977	        47.31 ns/op	      16 B/op	       1 allocs/op
```

### CODE REVIEW

**Comment by soujanyanmbri** [comment]
Before:
```
goos: darwin
goarch: arm64
pkg: github.com/corazawaf/coraza/v3/testing/coreruleset
cpu: Apple M1 Pro
BenchmarkCRSCompilation-10           163          72857088 ns/op        89749506 B/op     940950 allocs/op
BenchmarkCRSSimpleGET-10           15865            789790 ns/op          156790 B/op       3634 allocs/op
BenchmarkCRSSimplePOST-10           9046           1333304 ns/op          316277 B/op       6416 allocs/op
BenchmarkCRSLargePOST-10             122         100314446 ns/op          780956 B/op       6444 allocs/op
```

After: 
```
goos: darwin
goarch: arm64
pkg: github.com/corazawaf/coraza/v3/testing/coreruleset
cpu: Apple M1 Pro
BenchmarkCRSCompilation-10           172          70637522 ns/op        89748519 B/op     940951 allocs/op
BenchmarkCRSSimpleGET-10           16035            750342 ns/op          157150 B/op       3634 allocs/op
BenchmarkCRSSimplePOST-10           7284           1403836 ns/op          317016 B/op       6416 allocs/op
BenchmarkCRSLargePOST-10             100         100243078 ns/op          746964 B/op       6439 allocs/op
```

File: internal/collections/map.go
Comment (jcchavezs) [inline internal/collections/map.go:44]: Wonder why we removed this.

File: internal/collections/map.go
Comment (jcchavezs) [inline internal/collections/map.go:125]: Here we are losing case sensitivenes, do you mind explaining why?

File: internal/collections/named.go
Comment (jcchavezs) [inline internal/collections/named.go:68]: 👍

File: internal/collections/map_test.go
Comment (jcchavezs) [inline internal/collections/map_test.go:114]: I think the interpolation here impacts the benchmark. I would generate the keys in a map first.

File: internal/collections/map.go
Comment (soujanyanmbri) [reply internal/collections/map.go:44]: If c.data is empty, the c.data[key] lookup will return nil, so checking len(c.data) == 0 first wasn't required.

Comment (jcchavezs) [reply internal/collections/map.go:44]: true but we save the `strings.ToLower` if data is empty :)

File: internal/collections/map_test.go
Comment (soujanyanmbri) [reply internal/collections/map_test.go]: Fixed it 👍

File: internal/collections/map.go
Comment (soujanyanmbri) [reply internal/collections/map.go:44]: Ah, you're right. I'll add it back!

File: internal/collections/map.go
Comment (soujanyanmbri) [reply internal/collections/map.go:125]: I'm not entirely sure. That has been there since before, let me check why

Comment (soujanyanmbri) [reply internal/collections/map.go]: Oh just realised my mistake. Fixed that. Thank you!

**Comment by jcchavezs** [comment]
also please post the results of the benchmarks

**Comment by soujanyanmbri** [comment]
```
Before: 
BenchmarkTxSetGet/Set-10         	26881494	        42.25 ns/op	      32 B/op	       1 allocs/op
BenchmarkTxSetGet/Get-10         	15812412	        76.35 ns/op	      48 B/op	       2 allocs/op
After: 
BenchmarkTxSetGet/Set-10         	66049288	        18.02 ns/op	       0 B/op	       0 allocs/op
BenchmarkTxSetGet/Get-10         	25339977	        47.31 ns/op	      16 B/op	       1 allocs/op
```
This is after using the slice reuse 
CC: @jcchavezs

File: internal/collections/map_test.go
Comment (jcchavezs) [inline internal/collections/map_test.go:124]: since you already set the values before, shall we do you do this here again? also, likely we reset the benchmarks after this to not to pollute what we are measuring.

Comment (soujanyanmbri) [reply internal/collections/map_test.go]: Done, fixed it.

File: internal/collections/map_test.go
Comment (jcchavezs) [inline internal/collections/map_test.go:117]: Do we need this before the `Set` benchmark?

Comment (soujanyanmbri) [reply internal/collections/map_test.go]: Ah, you're right. was redundant, removed it.

**Review by M4tteoP — COMMENTED** [review]
Let's wait for @jcchavezs final approval, but LGTM, thanks!
Posting some more before/after benchmarks
Before:

```
BenchmarkCRSSimplePOST-10    	     872	   1312620 ns/op	  317136 B/op	    6415 allocs/op
BenchmarkCRSLargePOST-10    	      12	  97123340 ns/op	 1694983 B/op	    6646 allocs/op
```

After:

```
BenchmarkCRSSimplePOST-10    	     865	   1331177 ns/op	  316196 B/op	    6374 allocs/op
BenchmarkCRSLargePOST-10    	      12	  99717045 ns/op	 1492402 B/op	    6553 allocs/op
```

**Comment by jptosso** [comment]
lgtm

**Review by jcchavezs — APPROVED** [review]

**Comment by jcchavezs** [comment]
Thanks @soujanyanmbri !

**Pull request merged**
